tsl/src/remote/connection.c

```diff
@@ -1852,6 +1954,137 @@ remote_connection_set_single_row_mode(TSConnection *conn)
 	return PQsetSingleRowMode(conn->pg_conn);
 }
 
+static bool
+send_binary_copy_header(const TSConnection *conn, TSConnectionError *err)
+{
+	/* File header for binary format */
+	static const char file_header[] = {
+		'P', 'G', 'C', 'O', 'P', 'Y', '\n', '\377', '\r', '\n', '\0', /* Signature */
+		0,   0,   0,   0,											  /* 4 bytes flags */
+		0,   0,   0,   0 /* 4 bytes header extension length (unused) */
+	};
+
+	int res = PQputCopyData(conn->pg_conn, file_header, sizeof(file_header));
+
+	if (res != 1)
+		return fill_connection_error(err,
+									 ERRCODE_CONNECTION_FAILURE,
+									 "could not set binary COPY mode",
+									 conn);
+	return true;
+}
+
+bool
+remote_connection_begin_copy(TSConnection *conn, const char *copycmd, bool binary,
+							 TSConnectionError *err)
+{
+	PGconn *pg_conn = remote_connection_get_pg_conn(conn);
+	PGresult *volatile res = NULL;
+
+	if (PQisnonblocking(pg_conn))
+		return fill_simple_error(err,
+								 ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "distributed copy doesn't support non-blocking connections",
+								 conn);
+
+	if (conn->status != CONN_IDLE)
+		return fill_simple_error(err,
+								 ERRCODE_INTERNAL_ERROR,
+								 "connection not IDLE when beginning COPY",
+								 conn);
+
+	res = PQexec(pg_conn, copycmd);
+
+	if (PQresultStatus(res) != PGRES_COPY_IN)
+	{
+		fill_result_error(err,
+						  ERRCODE_CONNECTION_FAILURE,
+						  "unable to start remote COPY on data node",
+						  res);
+		PQclear(res);
+		return false;
+	}
+
+	PQclear(res);
+
+	if (binary && !send_binary_copy_header(conn, err))
```

Comment:
I am not 100% sure as I did not implement this (just moved and refactored the code).

But, I believe this might help explain: Basically, the `FORMAT BINARY` in the `COPY` statement has nothing to do with the wire protocol in `libpq`. It just tells the system the format of a "file" read by the server. Note that if you do the following statement:

```
COPY mytable FROM 'my_data_file` FORMAT BINARY;
```

you are basically telling the _server_ to open and read a file in binary format that must exist on the server. This is different from the `libpq` wire format when sending the data from the client to the server.

When we are sending data from a client, the COPY statement sets things up to read from STDIN instead of a file. In other words, the server does not read the file locally.

The copy header sent here is the format of the data sent with `libpq` from the client to the server and not file format.

Reply:
I would rather reference this one instead https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-COPY. I was curious why we had to do this hack and send raw protocol header instead of using libpq, which likely would do the same under the hood